feat(prune): timeout

emhane opened the pull request:
Closes https://github.com/paradigmxyz/reth/issues/6770

Adds a 100 ms time out for pruning.

Currently, it does not enforce the timeout for `AccountHistory` and `StorageHistory` prune segments. Will be fixed in the next PR.

shekhirin commented on crates/storage/db/src/abstraction/database.rs:
```diff
@@ -4,7 +4,10 @@ use crate::{
-use std::{fmt::Debug, sync::Arc};
+use std::{fmt::Debug, sync::Arc, time::Duration};
+
+/// Time limit that pruning can run before it's forced to yield the [`DatabaseProviderRW`] hook.
+pub const DEFAULT_TIME_OUT_DURATION_PRUNER: Duration = Duration::from_millis(100);
```
I think a cleaner way would be to introduce a struct containing both `delete_limit` representing the max number of items to delete and `time_limit` representing the max allowed time to spend on deleting, making both of them `Option`. And then pass this struct as an argument to `prune_table_with_range`.

We definitely want to control the pruning time outside of the database abstraction.

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -145,9 +151,10 @@ impl<DB: Database> Pruner<DB> {
 
         for (segment, purpose) in segments {
-            if delete_limit == 0 {
+            if limit.at_limit(start, deleted) {
```
we always pass the same `start`, can we make it a field of the `PruneLimit` struct? So only the `deleted` input argument will change

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -185,7 +192,7 @@ impl<DB: Database> Pruner<DB> {
 
                 done = done && output.done;
-                delete_limit = delete_limit.saturating_sub(output.pruned);
+                deleted += 1;
```
I think this is incorrect: `deleted` means here the number of pruned segments, but we use it as a number of deleted entries in the `limit.at_limit`.

I think we should rather adjust the `PruneLimit` deletion limit budget when we finish pruning the segment.

emhane replied:
on main, `delete_limit` only exists within scope of this function. I think the change that needs to happen is rename the `entries_limit` to `items_limit` to encompass both segments and entries. the 2 layers basically, where entries are nested in segments.

emhane replied:
now I see what you mean. my bad, I was really hungry.

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -2515,3 +2517,51 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
+#[derive(Debug, Default, Clone, Copy, Constructor)]
+pub struct PruneLimit {
+    /// Maximum entries to delete from the database in one prune job.
+    segment_limit: Option<usize>,
```
this shouldn't be a limit per segment, it's a limit for the number of entries that were deleted in one Pruner run

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -2515,3 +2517,51 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
+}
+
+impl PruneLimit {
+    /// Returns the maximum entries that can be deleted from the database in one prune job. `None`
+    /// is equivalent to unlimited segments.
```
I think there's confusion between entries and segments:
- Entries are basically rows in the database, and we don't want to delete more than X of them in one pruner run
- Segments are parts of the database that represent one or more tables, and we don't want to limit them as long as the number of deleted entries is respected

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -821,25 +823,25 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
-                    deleted += 1;
-                    delete_callback(row);
-                }
+        while let Some(row) = walker.next().transpose()? {
```
we should leave the initial limit check because otherwise we'll do a `walker.next()` that's not going to be used later

emhane replied:
seems to be a bug on main too, will open an issue

https://github.com/paradigmxyz/reth/blob/48b763f88a3cf175ded5d5ea9f99559a59c65087/crates/storage/provider/src/providers/database/provider.rs#L845

emhane replied:
wait, actually, must have been correct from the start, right? walker only lives in scope of this method, so shouldn't be a problem to skip limit check outside of loop - if limit is reached, then the method returns anyway.

shekhirin reviewed (changes requested):
I like the `PruneLimit` struct, it's the right direction.

We want it to control when to stop **one** `Pruner` run based on both:
1. Number of entries deleted from the database (total, from all segments)
2. Time spent running (again, for all segments, and to be more precise for the whole `Pruner::run`)

emhane commented:
> I like the `PruneLimit` struct, it's the right direction.
> 
> We want it to control when to stop **one** `Pruner` run based on both:
> 
> 1. Number of entries deleted from the database (total, from all segments)
> 2. Time spent running (again, for all segments, and to be more precise for the whole `Pruner::run`)

yes nice we had the same idea :)

emhane commented:
commit [1e43672](https://github.com/paradigmxyz/reth/pull/6958/commits/1e43672413426171def4a43302a9f4a3efef15e0) wasn't really necessary since it's looping through a non-owning iterator. one test is still failing, will check it later.

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -146,69 +180,76 @@ impl<DB: Database> Pruner<DB> {
+                    debug!(target: "pruner", segment = ?segment.segment(), ?purpose, "Nothing to prune for the segment");
+                }
+
+                if limiter.at_limit() {
+                    done = false;
+                    break
```
unsure if that's correct, what if we actually finished (`done = true`), but also hit the limit?

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -146,69 +180,76 @@ impl<DB: Database> Pruner<DB> {
                     );
+
+                    if output.pruned > 0 {
+                        // sets `is_timed_out`
+                        let _ = limiter.at_limit();
```
imo this is a hidden behavior we want to avoid

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -2515,3 +2518,93 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
+    pub fn new_with_fraction_of_units_limit(old: Self, denominator: NonZeroUsize) -> Self {
+        let Self { deleted_units_limit, job_timeout, start, .. } = old;
+
+        Self::new(deleted_units_limit.map(|limit| limit / denominator), job_timeout, start)
```
how do you feel about making it a builder pattern, so it'd be a `with_fraction_of_units_limit(mut self, fraction: NonZero) -> Self` to just replace one field?

emhane replied:
builder pattern should be reserved for builders, so added a builder now, took some time but looks clean

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -2515,3 +2518,93 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
+    }
+
+    /// Returns true if prune limit is reached.
+    pub fn at_limit(&mut self) -> bool {
```
nit: because it returns a boolean, but it's likely a style preference
```suggestion
    pub fn is_limit_reached(&mut self) -> bool {
```

shekhirin commented on crates/storage/provider/src/providers/database/provider.rs:
```diff
@@ -2515,3 +2518,93 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
+    pub fn at_limit(&mut self) -> bool {
+        let Self { deleted_units_limit, deleted_units_count, job_timeout, start, .. } = self;
+
+        if let Some(limit) = deleted_units_limit {
+            if limit == deleted_units_count {
+                return true
+            }
+        }
+        if let Some(timeout) = job_timeout {
+            if *timeout <= start.elapsed() {
+                self.timed_out = true;
+                return true
+            }
+        }
```
can we make it consistent between two reasons, so that we set the reason flag inside the limiter?

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -265,6 +306,7 @@ impl<DB: Database> Pruner<DB> {
 
 #[cfg(test)]
 mod tests {
```
test nit: it'd be ideal to test the limiter logic by introducing a custom `TestSegment: Segment` that either hits the units deletion limit, or the timeout.

emhane replied:
will save this for last :)

emhane replied:
maybe out of scope? since storage and account segments are bundled into one type atm in test db, improving the test type system further can be an issue with both these tasks, wdyt?

shekhirin reviewed (changes requested):
Nice, it basically works and does what it should do: stops the pruner when we either deleted max allowed number of units per run or hit the timeout.

shekhirin commented on crates/prune/src/segments/headers.rs:
```diff
@@ -44,21 +44,27 @@ impl<DB: Database> Segment<DB> for Headers {
+        // todo: divide time fairly between all three
         let results = [
-            self.prune_table::<DB, tables::Headers>(provider, block_range.clone(), delete_limit)?,
+            self.prune_table::<DB, tables::Headers>(provider, block_range.clone(), limiter)?,
             self.prune_table::<DB, tables::HeaderTerminalDifficulties>(
                 provider,
                 block_range.clone(),
-                delete_limit,
+                limiter,
             )?,
-            self.prune_table::<DB, tables::CanonicalHeaders>(provider, block_range, delete_limit)?,
+            self.prune_table::<DB, tables::CanonicalHeaders>(provider, block_range, limiter)?,
```
yeah, agreed. I think the ideal solution would be to open three cursors and walk all of them at the same time, deleting the entries. And once the time limit is hit, stop.

But it's hard to fit in the `prune_table` method, we should probably do it some other way here.

emhane commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +91,90 @@ impl ReceiptsLogPruneConfig {
+    /// [PruneProgress::HasMoreData] is returned.
+    pub fn new(done: bool, timeout: bool) -> Self {
         if done {
             Self::Finished
+        } else if timeout {
+            Self::timed_out()
         } else {
-            Self::HasMoreData
+            Self::segment_limit_reached()
+        }
```
in tests it gets a little messy if I init variants directly, but will add `new` to method name

emhane commented:
> Nice, it basically works and does what it should do: stops the pruner when we either deleted max allowed number of units per run or hit the timeout.

not completely done addressing review, to do so need clarification on 'deleted max allowed number of units'. per prune job or per block? first I though we would do 'per call to prune', but then I got 'block' from here

https://github.com/paradigmxyz/reth/blob/3c028e5ef401d7f1f4d45b182bbfa74b5228a0b6/crates/prune/src/builder.rs#L48-L52

emhane commented on crates/prune/src/pruner.rs:
```diff
@@ -135,7 +168,7 @@ impl<DB: Database> Pruner<DB> {
+        mut limiter: PruneLimiter,
     ) -> Result<(PrunerStats, usize, PruneProgress), PrunerError> {
         let static_file_segments = self.static_file_segments();
         let segments = static_file_segments
```
we should make this a ring, so the last pruned segment is first on next call. it should already have happened without the timeout. since otherwise we risk pruning the transactions but not the headers and less likely the receipts. @shekhirin , thoughts?

shekhirin replied:
agreed, that makes sense. Let's do it in a different PR

emhane replied:
https://github.com/paradigmxyz/reth/issues/7343

shekhirin commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +91,76 @@ impl ReceiptsLogPruneConfig {
+    /// Prune job timed out.
+    Timeout,
+    /// Limit on the number of deleted entries (rows in the database) per prune job was reached.
+    LimitEntriesDeleted,
```
maybe something like that?
```suggestion
    DeletedEntriesLimitReached,
```

shekhirin commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +91,76 @@ impl ReceiptsLogPruneConfig {
-    /// If `done == true`, returns [PruneProgress::Finished], otherwise [PruneProgress::HasMoreData]
-    /// is returned.
-    pub fn from_done(done: bool) -> Self {
+    /// If `done == true`, returns [`PruneProgress::Finished`], otherwise
+    /// [PruneProgress::HasMoreData] is returned.
```
nit: consistency between `[]` and `[``]` usage
```suggestion
    /// If `done == true`, returns [PruneProgress::Finished], otherwise
    /// [PruneProgress::HasMoreData] is returned.
```

shekhirin commented on crates/prune/src/segments/headers.rs:
```diff
@@ -159,11 +255,17 @@ mod tests {
             let provider = db.factory.provider_rw().unwrap();
-            let result = segment.prune(&provider, input).unwrap();
+            let result = segment.prune(&provider, input.clone()).unwrap();
+            trace!(target: "pruner::test",
+                expected_prune_progress=?expected_result.0,
+                expected_pruned=?expected_result.1,
+                result=?result,
+                "PruneOutput"
+            );
```
do we need it in tests?

emhane replied:
yes, it's needed. when the assertion fails, the fields of the expected `PruneOutput` aren't printed.

shekhirin commented on crates/prune/src/pruner.rs:
```diff
@@ -146,69 +179,82 @@ impl<DB: Database> Pruner<DB> {
+        Ok((
+            stats,
+            limiter.deleted_entries_count(),
+            PruneProgress::new(done, limiter.is_timed_out()),
```
I think this API is confusing. Can we just construct the enum manually? Because passing the `timeout: bool` and falling back to the entries limit as a default seems easy to misuse

shekhirin reviewed (commented):
Will finish the review once tests are passing and it's ready

emhane commented on crates/prune/src/segments/history.rs:
```diff
@@ -34,79 +33,180 @@ where
+            "cannot prune complete shard"
+        );
+
         // Shard contains block numbers that are higher than the target one, so we need to
         // filter it. It is guaranteed that further shards for this sharded key will not
         // contain the target block number, as it's in this shard.
```
```suggestion
        // Shard contains block numbers that are higher than the target one, so we need to
        // filter it. It is guaranteed that further shards for this sharded key will not
        // contain the target block number, if it's in this shard.
```

@shekhirin ? shard may contain block numbers that are only higher than the target one too.

shekhirin replied:
> shard may contain block numbers that are only higher than the target one too.

right, true

emhane commented on crates/prune/src/segments/history.rs:
```diff
@@ -34,79 +33,180 @@ where
+
+        // If there were blocks less than or equal to the target one
+        // (so the shard has changed), update the shard.
+        if blocks.len() as usize != old_blocks.len() {
+            // If there are no more blocks in this shard, we need to remove it, as empty
+            // shards are not allowed.
+            if old_blocks.is_empty() {
```
this seems like a bug, there is still blocks in the shard, since otherwise the check `if blocks.len() as usize != old_blocks.len()` wouldn't pass @shekhirin will open an issue

shekhirin replied:
why? if `blocks = [1, 2, 3]`, `to_block = 3`, then `old_blocks = []` and the check will pass `blocks.len() != old_blocks.len()`. Then, we'll go into this clause `old_blocks.is_empty()` and remove the shard completely.

emhane replied:
https://github.com/paradigmxyz/reth/issues/7104

emhane replied:
yeah exactly, so there are still blocks `[1, 2, 3]` in the shard. old blocks helped me think about it: lower block number = older block

emhane replied:
shard should only be completely removed if there is guarantee all blocks are lower than target block. which is already done here

https://github.com/paradigmxyz/reth/blob/e4f34e79ab144734d38661b66e3ef3e32b9a154d/crates/prune/src/segments/history.rs#L40-L42

emhane replied:
renamed `old_blocks` to `lower_blocks`, to stick to standard concept of 'block height'

emhane replied:
fudge, `higher_blocks` it should be

shekhirin replied:
sorry, I needed to clarify my example. If it's **the last shard** for this sharded key, then the `if key.as_ref().highest_block_number <= to_block {` check will not pass, because the `highest_block_number == u64::MAX`. So we go into this clause, filter the blocks and realize the shard is empty now, and delete it.

But I think this clause https://github.com/paradigmxyz/reth/blob/d6aea5f99da9ff85aaaddc8a426313123929bd7c/crates/prune/src/segments/history.rs#L167-L176 is actually redundant, it shouldn't be possible to get into it. I think I put it here jus to be sure all if-else branches are considered.

shekhirin commented on crates/prune/src/segments/headers.rs:
```diff
@@ -36,90 +36,174 @@ impl<DB: Database> Segment<DB> for Headers {
+        // todo: move this check to prune job start where limiter is made and update tests
+        // accordingly
+        if limiter.is_limit_reached() {
+            // Nothing to do, `input.delete_limit` is less than 3 so we can't prune all
```
why are we sure it's not because of the timeout?

shekhirin commented on crates/prune/src/segments/headers.rs:
```diff
@@ -36,90 +36,174 @@ impl<DB: Database> Segment<DB> for Headers {
+        let mut last_pruned_block_canonical = None;
+        // todo: guarantee skip filter and delete callback are same for all header table types
+
+        if let Err(err) =
+            provider.with_walker::<tables::Headers, _, _>(block_step.clone(), |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_headers = Some(row.0)
+                })
+            })
+        {
+            return Some(Err(err.into()))
+        }
+
+        if let Err(err) = provider.with_walker::<tables::HeaderTerminalDifficulties, _, _>(
+            block_step.clone(),
+            |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_td = Some(row.0)
+                })
+            },
+        ) {
+            return Some(Err(err.into()))
+        }
+
+        if let Err(err) =
+            provider.with_walker::<tables::CanonicalHeaders, _, _>(block_step, |ref mut walker| {
+                provider.step_prune_range(walker, limiter, &mut |_| false, &mut |row| {
+                    last_pruned_block_canonical = Some(row.0)
+                })
+            })
+        {
+            return Some(Err(err.into()))
+        }
```
as far as I understand, it means that we open a new cursor on every iteration. It's not good and adds a significant overhead, we should rather initialize the cursors once on `::new(...)` and then advance them on each iteration.

shekhirin commented on crates/prune/src/segments/history.rs:
```diff
@@ -34,79 +34,412 @@ where
     // 2. If the shard has `highest_block_number` greater than the target block number for
     // pruning, filter block numbers inside the shard which are less than the target
     // block number for pruning.
```
should we move these comments to the `step_prune_indices` function?

shekhirin commented on crates/prune/src/segments/history.rs:
```diff
@@ -7,15 +7,16 @@ use reth_db::{
 ///
 /// Returns total number of processed (walked) and deleted entities.
-pub(crate) fn prune_history_indices<DB, T, SK>(
+pub(crate) fn _prune_history_indices<DB, T, SK>(
```
`_` 🤔

shekhirin commented on crates/prune/src/segments/account_history.rs:
```diff
@@ -35,208 +37,341 @@ impl<DB: Database> Segment<DB> for AccountHistory {
+        };
+
+        if let Err(err) = provider.with_cursor::<tables::AccountsHistory, _, _>(|ref mut cursor| {
+            step_prune_indices::<DB, _, _>(
+                cursor,
+                to_block,
+                limiter,
+                &|a, b| a.key == b.key,
+                &|key| ShardedKey::last(key.key),
+            )
```
if I understand correctly, what happens here is that we prune one block in `AccountChangeSets`, and then delete this one block from all indices. If that's so, I think it'd be very slow because indices tables are large.

I'm not sure yet how to correctly solve this considering the timeout.

shekhirin commented on crates/prune/src/segments/mod.rs:
```diff
@@ -57,16 +63,42 @@ pub trait Segment<DB: Database>: Debug + Send + Sync {
+    pub(crate) limiter: PruneLimiter,
+}
+
+impl Clone for PruneInput {
+    fn clone(&self) -> Self {
+        let Self { previous_checkpoint, to_block, limiter } = self;
+
+        let previous_checkpoint = *previous_checkpoint;
+        let to_block = *to_block;
+        let limiter = PruneLimiterBuilder::floor_deleted_entries_limit_to_multiple_of(
+            limiter,
+            NonZeroUsize::new(1).unwrap(),
+        )
+        .build();
+
+        Self { previous_checkpoint, to_block, limiter }
+    }
```
IMO we shouldn't hide this behaviour inside the clone, and it should be done explicitly

shekhirin commented on crates/prune/src/segments/mod.rs:
```diff
@@ -125,29 +157,39 @@ impl PruneInput {
+            .unwrap_or(0)
+    }
+
+    pub(crate) fn _get_last_pruned_block(&self) -> Option<u64> {
```
do we need this?

shekhirin commented on crates/storage/db/src/tables/models/accounts.rs:
```diff
@@ -73,6 +73,11 @@ impl BlockNumberAddress {
 
+    /// Create a new [`RangeInclusive`] from `start` to `end`
+    pub fn range_inclusive(range: RangeInclusive<BlockNumber>) -> RangeInclusive<Self> {
+        (*range.start(), Address::ZERO).into()..=(*range.end(), Address::ZERO).into()
```
I think it won't include the `(*range.end, Address::MAX)`

shekhirin reviewed (changes requested):
Overall, it's the right direction and should work, good job and thanks for taking a deep dive into the pruner code and making the test suite better.

What concerns me is the significant slowdown in segments where we need to prune more than one table (headers, transactions, accounts/storage history):
1. We create a cursor on every deleted row
2. We walk through all account/storage indices on every deleted block in changesets

I propose to:
1. Fix the cursors, so that they're initialized only once and then reused for the duration of segment pruning
2. Make the account/storage history segments unaware of timeout, and only enforce the deletion limit on them. I think it should be in a separate PR, because the logic is (already was) very complex, it's easy to get wrong and we already know at least of one bug associated with it (when the indices contain blocks that are already pruned from the changesets).

shekhirin commented:
> cursor takes a rw lock on provider, so that isn't possible, afaik.

that should work, see Bodies stage https://github.com/paradigmxyz/reth/blob/553a27153398d12973c65ebb3ebdc411e65e0484/crates/stages/src/stages/bodies.rs#L117-L122

emhane commented on crates/prune/src/builder.rs:
```diff
@@ -51,6 +58,12 @@ impl PrunerBuilder {
         self
     }
 
+    /// Sets the timeout for pruner, per job.
```
```suggestion
    /// Sets the timeout for pruner, per job. Caution! For a full node, this time out is a soft limit, in the sense that 
    /// history pruning will prune the given range before checking the timeout, unlike headers which prunes step by 
    /// step and checks the timeout on each step.
```

emhane commented on crates/prune/src/pruner.rs:
```diff
@@ -130,25 +158,25 @@ impl<DB: Database> Pruner<DB> {
     /// Prunes the segments that the [Pruner] was initialized with, and the segments that needs to
     /// be pruned according to the highest static_files.
```
```suggestion
    /// Prunes the segments that the [Pruner] was initialized with, and the segments that needs to
    /// be pruned according to the highest static_files. Segments are parts of the database that represent one or 
    /// more tables.
```
copied that text from one of your previous review comments, it's very helpful

emhane reviewed (commented):
best team work, high-five! looks very good

shekhirin commented on crates/primitives/src/prune/limiter.rs:
```diff
@@ -0,0 +1,123 @@
+        self
+    }
+
+    /// Returns `true` if time limit is reached.
+    pub fn is_time_limit_reached(&self) -> bool {
+        self.time_limit.as_ref().map_or(false, |limit| limit.is_limit_reached())
+    }
```
hmm why the false positive?

emhane replied:
because if pruning exists because of some other reason, i.e. `self.start.elapsed() >= self.limit` is false, and then some more work is done, **then** the check is run again, it might be true the second time as more time will have elapsed.

shekhirin replied:
> because if pruning exists because of some other reason, i.e. self.start.elapsed() >= self.limit is false, and then some more work is done, then the check is run again, it might be true the second time as more time will have elapsed.

I don't understand why it matters; if the check has already returned "true" on the first call (due to entries limit), so on the second call even though another limit (time) has been reached, it still returns "true"?

emhane replied:
yeah but if it returns false the first time, and then true, you have a false positive

emhane replied:
I'm talking ab if someone in the future decides to use that method on the limiter to check if pruning timed out

shekhirin commented on crates/primitives/src/prune/limiter.rs:
```diff
@@ -0,0 +1,123 @@
+    time_limit: Option<PruneTimeLimit>,
+}
+
+#[derive(Debug, Clone)]
+struct PruneDeletedEntriesLimit {
+    /// Maximum entries (rows in the database) to delete from the database.
+    limit: usize,
+    /// Current number of entries (rows in the database) that have been deleted.
+    deleted: usize,
+}
+
+impl PruneDeletedEntriesLimit {
+    fn new(limit: usize) -> Self {
+        Self { limit, deleted: 0 }
+    }
+
+    fn is_limit_reached(&self) -> bool {
+        self.deleted >= self.limit
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PruneTimeLimit {
+    /// Maximum duration of one prune run.
+    limit: Duration,
+    /// Time when the prune run has started.
+    start: Instant,
+}
+
+impl PruneTimeLimit {
+    fn new(limit: Duration) -> Self {
+        Self { limit, start: Instant::now() }
+    }
+
+    fn is_limit_reached(&self) -> bool {
+        self.start.elapsed() >= self.limit
+    }
+}
```
unsure about it, it makes the struct definitions located too far from their usages in `PruneLimiter`

emhane replied:
yeah ok, nice, have always been told to do it that way, now I shall also deviate from that rule 😏

shekhirin commented on crates/primitives/src/prune/limiter.rs:
```diff
@@ -0,0 +1,123 @@
+        Self { limit, deleted: 0 }
+    }
+
+    fn is_limit_reached(&self) -> bool {
+        self.deleted >= self.limit
+    }
```
not sure if we want to generalize this behavior and hide the limiters behind traits, there's only two of them and we don't expect to add more. Think we can add this later if needed (so if another type of limit will be introduced)

mattsse commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +93,61 @@ impl ReceiptsLogPruneConfig {
+    }
+
+    /// Returns `true` if the reason is timeout.
+    pub fn is_timeout(&self) -> bool {
```
```suggestion
    pub const fn is_timeout(&self) -> bool {
```

mattsse commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +93,61 @@ impl ReceiptsLogPruneConfig {
+    }
+
+    /// Returns `true` if the reason is reaching the limit on deleted entries.
+    pub fn is_entries_limit_reached(&self) -> bool {
```
```suggestion
    pub const fn is_entries_limit_reached(&self) -> bool {
```

mattsse commented on crates/primitives/src/prune/mod.rs:
```diff
@@ -91,21 +93,61 @@ impl ReceiptsLogPruneConfig {
     }
+
+    /// Returns `true` if prune run is finished.
+    pub fn is_finished(&self) -> bool {
```
```suggestion
    pub const fn is_finished(&self) -> bool {
```

mattsse commented on crates/prune/src/pruner.rs:
```diff
@@ -52,6 +59,7 @@ impl<DB: Database> Pruner<DB> {
         min_block_interval: usize,
         delete_limit: usize,
```
rename to `delete_limit_per_block`?

mattsse commented on crates/primitives/src/prune/limiter.rs:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    fn is_limit_reached(&self) -> bool {
+        self.start.elapsed() >= self.limit
```
```suggestion
        self.start.elapsed() > self.limit
```

shekhirin replied:
🤔 why?

mattsse reviewed (changes requested):
this setup is quite easy to follow.

only have a few nits and would appreciate another look by @joshieDo

shekhirin commented on crates/prune/src/segments/headers.rs:
```diff
@@ -36,90 +42,164 @@ impl<DB: Database> Segment<DB> for Headers {
-            // headers-related tables up to the same height
-            return Ok(PruneOutput::not_done())
-        }
+        let last_pruned_block =
+            if block_range_start == 0 { None } else { Some(block_range_start - 1) };
```
yep, think we can do it in another PR, but agreed

joshieDo commented on crates/prune/src/builder.rs:
```diff
@@ -51,6 +58,15 @@ impl PrunerBuilder {
 
+    /// Sets the timeout for pruner, per run.
+    ///
+    /// CAUTION: Account and Storage History prune segments treat this timeout as a soft limit,
```
Why?

shekhirin replied:
we do account/storage history pruning in two steps:
1. Prune the changesets and memorize the highest pruned block
2. Prune indices up to the highest pruned block from changesets

Given that, we can't stop the step 2 if time limit is reached, because then we'll have inconsistent data in the database (partly pruned block).

shekhirin replied:
It needs to be solved separately because not obvious yet how to do that and the PR is already large

joshieDo commented on crates/prune/src/segments/account_history.rs:
```diff
@@ -5,10 +5,13 @@ use crate::{
 use tracing::{instrument, trace};
 
+/// Number of account history tables to prune in one step
+const ACCOUNT_HISTORY_TABLES_TO_PRUNE: usize = 2;
```
`Tables` naming makes lil sense to me here

shekhirin replied:
improved the comment

joshieDo reviewed (approved):
lgtm, but would feel more confortable if it was run for some time if it hasn't already
